r_code/ensemble_estimator.R

```diff
@@ -0,0 +1,356 @@
+  # Define the estimator call, and apply.
+  # MC and SDID Unconstrained are exceptions to the rule (no unique function).
+  if (method_name == "MC") {
+    return(EstimateGsynthSeries(placebo_data, estimator = "mc", ...))
+  }
+  if (method_name == "SDID_Uncon") {
+    return(EstimateSDIDSeries(placebo_data, constrained = F, ...))
+  }
+  # Define the call to the function based on method name, and cal.
+  estimator_call <- paste("Estimate", method_name, "Series", sep = "")
+  est_out <- do.call(estimator_call, list(placebo_data, ...))
+  return(est_out)
+}
+#' Estimates ensemble outcomes combining methods with placebo optimal weights.
+#'
+#' @param method_names Vector of strings for the names of methods to ensemble.
+#'    Names must be one of the potential methods, will be used to call
+#'    Estimate"method_names"Series.
+#' @param true_data Tibble of the formatted raw data to estimate ensemble for.
+#'    This should be a true panel in that every unit has an observed outcome
+#'    for every time period in the data set -- remove NAs beforehand.
+#' @param pred_list List of tibbles containing the estimated series per method.
+#' @param constrained Boolean for whether to constrain the ensemble to have
+#'    non-negative weights that sum to one over the input methods.
+#' @param intercept_allowed Boolean for whether to include an intercept
+#'    adjustment in the ensemble.
+#' @param indiv_weights Boolean for whether to estimate unit-specific weights.
+#' @param time_var String name of the time variable in the raw data.
+#' @param id_var String name of the unit variable in the raw data.
+#' @param outcome_var String name of the outcome variable in the raw data.
+#' @param treat_indicator String name of treatment indicator in the raw data.
+#' @param counterfac_var String name of the counterfactual variable in the raw
+#'    data. Can be NULL, as the placebo creates it's own counterfactual.
+#'
+#' @return A long form tibble with the predictions from each method, their
+#'    associated weights, and the final ensembled point estimate as well as the
+#'    effects.
+# TODO(alexdkellogg): Potential improvement - focus ensemble weights on k
+#    periods after treatment, rather than the full post-treat time period.
+# TODO(alexdkellogg): Update the ensemble so that it averages over several
+#    placebo datasets. Learn time vs accuracy trade-off.
+EstimateEnsemble <- function(method_names, true_data, pred_list,
+                             constrained = T, intercept_allowed = T,
+                             indiv_weights = F, time_var = "period",
+                             id_var = "entry", outcome_var = "target",
+                             treat_indicator = "treatperiod_0",
+                             counterfac_var = "counter_factual") {
+  # Ensure the method names are appropriately specified.
+  stopifnot(method_names %in%
+    c("CausalImpact", "Gsynth", "SDID", "SDID_Uncon", "SCM"))
```

Comment:
Do you want to use `match.arg` to ensure this?

Reply:
I kept as is because I was having issues with partial matching. I did add this statement in each of the "leaves" where it's used.